src/solver.jl

```diff
@@ -0,0 +1,682 @@
+
+            deleteat!(unchecked_solutions, to_be_removed)
+        end
+    catch
+        @warn "unable to verify solutions"
+    end
+
+    if length(unchecked_solutions) == 1
+        unchecked_solutions = unchecked_solutions[1]
+    end
+
+    return unchecked_solutions
+end
+
+#returns a dictionary of the solutions to the system of equations
+function solve_system_eq(equs::Vector{Equation}, vars)
+    removed = Equation[]#keep track of removed equations
+
+    reduced = copy(equs)#the reducing set
+
+    for i = 1:length(vars)#go through each variable and remove
+        remove_eq(reduced, vars[i], removed)
+    end
+
+    #solve last equation in the reduced set
+
+    solutions = Dict()
+
+    #re subsititute the variables back in to find value
+    for i = length(removed):-1:1
+        current_eq = substitute(removed[i], solutions)
+        solutions[current_eq.lhs] = current_eq.rhs
+    end
+
+    return solutions
+end
+
+#  =======  MAIN FUNCTIONS END ======
+
+function get_parts_list(a, b, a_list = Vector{Any}(), b_list = Vector{Any}())
+    if a isa SymbolicUtils.Sym
+        push!(a_list, a)
+        push!(b_list, b)
+    elseif istree(a) && istree(b) && isequal(operation(a), operation(b))
+        a_args = arguments(a)
+        b_args = arguments(b)
+
+        length(a_args) != length(b_args) && return Nothing
+
+        for i = 1:length(a_args)
+            check = get_parts_list(a_args[i], b_args[i], a_list, b_list)
+            check == Nothing && return Nothing
+        end
+    elseif a isa Equation && b isa Equation
+        get_parts_list(a.lhs, b.lhs, a_list, b_list)
+        get_parts_list(a.rhs, b.rhs, a_list, b_list)
+    else
+        return Nothing
+    end
+    return (a_list, b_list)
+end
+
+function find_matches(a)
+    vars_seen = UInt[]
+    matching = Vector{Int}[]
+
+    for i = 1:length(a)
+        match_set = Int[]
+
+        push!(match_set, i)
+
+        current = a[i]
+        (hash(current) in vars_seen) && continue
+
+        for j = i+1:length(a)
+            other = a[j]
+            if isequal(current, other)
+                push!(match_set, j)
+            end
+        end
+        push!(vars_seen, hash(current))
+
+        if length(match_set) > 1
+            push!(matching, match_set)
+        end
+
+    end
+    return matching
+end
+
+function verify_matches(a_parts, b_parts)
+    a_matches = find_matches(a_parts)
+    for match in a_matches
+        to_compare = b_parts[match[1]]
+        for i = 2:length(match)
+            !isequal(to_compare, b_parts[match[i]]) && return false
+        end
+    end
+    return true
+end
+
+struct Becomes
+    before::Any
+    after::Any
+end
+
+function create_eq_pairs(a, b)
+    out = Dict{SymbolicUtils.Sym,Any}()
+
+    for i = 1:length(a)
+        a_part = a[i]
+        a_part in keys(out) && continue
+
+        out[a_part] = b[i]
+    end
+
+    return out
+end
+
+function replace(expr, dic::Dict)
+    if expr isa SymbolicUtils.Sym && haskey(dic, expr)
```

Comment:
Could you always use `issym`, `isterm`, `ispow`, etc?

Reply:
updated! thanks